server/src/main/java/org/elasticsearch/common/util/concurrent/KeyedLock.java

```diff
@@ -95,38 +96,62 @@ public boolean isHeldByCurrentThread(T key) {
 
     private void release(T key, KeyLock lock) {
         assert lock == map.get(key);
-        final int decrementAndGet = lock.count.decrementAndGet();
+        final int decrementAndGet = lock.decCountAndGet();
         lock.unlock();
         if (decrementAndGet == 0) {
             map.remove(key, lock);
         }
         assert decrementAndGet >= 0 : decrementAndGet + " must be >= 0 but wasn't";
     }
 
-    private final class ReleasableLock extends AtomicBoolean implements Releasable {
-        final T key;
+    private final class ReleasableLock extends AtomicReference<T> implements Releasable {
         final KeyLock lock;
 
         private ReleasableLock(T key, KeyLock lock) {
-            this.key = key;
+            super(key);
             this.lock = lock;
         }
 
         @Override
         public void close() {
-            if (compareAndSet(false, true)) {
-                release(key, lock);
+            T k = getAndSet(null);
+            if (k != null) {
+                release(k, lock);
             }
         }
     }
 
-    @SuppressWarnings("serial")
     private static final class KeyLock extends ReentrantLock {
+        private static final VarHandle VH_COUNT_FIELD;
+
+        static {
+            try {
+                VH_COUNT_FIELD = MethodHandles.lookup().in(KeyLock.class).findVarHandle(KeyLock.class, "count", int.class);
+            } catch (NoSuchFieldException | IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        @SuppressWarnings("FieldMayBeFinal") // updated via VH_COUNT_FIELD (and _only_ via VH_COUNT_FIELD)
+        private volatile int count = 1;
+
         KeyLock() {
             super();
         }
 
-        private final AtomicInteger count = new AtomicInteger(1);
+        int decCountAndGet() {
+            do {
+                int i = count;
+                int newCount = i - 1;
```

Comment:
If you move the declaration of newCount outside the loop, then the loop condition can be the call to weakCompareAndSet instead of an infinite loop.

Reply:
You mean:

```java
            int newCount;
            int oldCount;
            do {
                oldCount = count;
                newCount = oldCount - 1;
            } while (VH_COUNT_FIELD.weakCompareAndSet(this, oldCount, newCount) == false);
            return newCount;
```

? I need `oldCount` outside of the loop as well. Not sure that's better? It's the same amount of lines and breaks with the style we used for this in `AstractRefCounted`?